podpac/core/compositor.py

```diff
@@ -229,35 +261,22 @@ def iteroutputs(self, coordinates):
             n_threads = 0
 
         if settings["MULTITHREADING"] and n_threads > 1:
-            # TODO pool of pre-allocated scratch space
-            # TODO: docstring?
-            def f(src):
-                return src.eval(coordinates)
-
-            # Create pool of size n_threads, note, this may be created from a sub-thread (i.e. not the main thread)
+            # evaluate nodes in parallel using thread pool
+            self._multi_threaded = True
             pool = thread_manager.get_thread_pool(processes=n_threads)
-
-            # Evaluate nodes in parallel/asynchronously
-            results = [pool.apply_async(f, [src]) for src in src_subset]
-
-            # Yield results as they are being requested, blocking when the thread is not finished
-            for src, res in zip(src_subset, results):
-                yield res.get()
-                # src._output = None # free up memory
-
-            # This prevents any more tasks from being submitted to the pool, and will close the workers one done
+            outputs = pool.map(lambda src: src.eval(coordinates), src_subset)
             pool.close()
-
-            # Release these number of threads back to the thread pool
             thread_manager.release_n_threads(n_threads)
-            self._multi_threaded = True
+            for output in outputs:
+                yield output
```

Comment:
@mpu-creare This is definitely a bug. When the `OrderedCompositor.composite` method breaks early (because there are no more NaN values to fill), the cleanup code after the yield loop never gets called. In this case, that includes setting the `_multi_threaded` attribute, closing the pool, and most importantly calling releasing the threads in the thread manager. The fix here evaluates the sources in parallel but waits for them all to finish before yielding results so that it can cleanup first.

The reason I've written all this is because there is a slight variation that we can consider. The following code still finishes evaluating all of the sources and cleans up the threads, but it also yields results as soon as they are ready. I believe this could be *marginally* more efficient, but I didn't think it was worth the unusual `GeneratorExit` idiom. I want to know if you think the following is preferable:

```
# evaluate nodes asynchronously using thread pool
pool = thread_manager.get_thread_pool(processes=n_threads)
results = [pool.apply_async(src.eval, [coordinates]) for src in src_subset]
pool.close()
try:
    for result in results:
        yield result.get()
except GeneratorExit:
    # the consuming generator was exited early
    # we still need to wait for the remaining threads and then release them from the thread manager
    for result in results:
        result.get()
thread_manager.release_n_threads(n_threads)
```

Note that it is correct to put the `thread_manager.release_n_threads(n_threads)` call *outside* the GeneratorExit clause because then it will be called whether the consuming loop breaks early or not.

Reply:
I think stick with what you have.